tmva/sofie/inc/TMVA/ROperator_Reduce.hxx

```diff
@@ -0,0 +1,301 @@
+      return input;
+   }
+
+   // shape of output tensors given input tensors
+   std::vector<std::vector<size_t>> ShapeInference(std::vector<std::vector<size_t>> input){
+      
+      // std::vector<std::vector<size_t>> ret;
+      // auto & input_shape = input[0];
+      // auto ret = std::vector<std::vector<size_t>>(1, input[0]); // return vector size 1 with first input
+      // return ret;
+      auto ret = input; //suggest copy to compiler
+      ret[fAttrAxes] = 1;
+      return ret;
+   }
+    void Initialize(RModel& model){
+
+        fUseSession = model.UseSession();
+
+        if (model.CheckIfTensorAlreadyExist(fNX) == false){   //input must be a graph input, or already initialized intermediate tensor
+            throw std::runtime_error("TMVA SOFIE Reduce Op Input Tensor " + fNX + " is not found in model");
+        }
+        fShapeX = model.GetTensorShape(fNX);
+        // find shape of Y and add it in the list of intermediate tensors
+        fShapeY = ShapeInference(fShapeX);
+        model.AddIntermediateTensor(fNY, model.GetTensorType(fNX), fShapeY);
+    }
+
+    std::string Generate(std::string OpName){
+      OpName = "op_" + OpName;
+      if (fShapeX.empty() || fShapeY.empty()) {
+         throw std::runtime_error("TMVA SOFIE Reduce Op called to Generate without being initialized first");
+      }
+
+      size_t outputLength = TMVA::Experimental::SOFIE::ConvertShapeToLength(fShapeY);
+
+      auto inputStrides = TMVA::Experimental::SOFIE::UTILITY::ComputeStrideFromShape(fShapeX);
+      auto outputStrides = TMVA::Experimental::SOFIE::UTILITY::ComputeStrideFromShape(fShapeY);
+      
+   // write here according to size of shape
+   // in generation code can be done automatically
+   // i0 =  i / s0 ; i1 = (i % s0) / s1 ; i2 = ( (i % s0) % s1 ) / s2 and so on
+   // and we have for the inverse
+   // i = i0 * s0 + i1 * s1 + i2 * s2 + i3 * s3 ....
+
+   // don't need to divide by last stride s[n-1] since it is 1 by definition
+      
+      std::stringstream out;
+      out << "\n//----  operator " << std::string(ReduceOperatorTrait<T,Op>::Name()) << "  " << OpName << "\n";
+      out << SP << "size_t dim = " << fShapeY.size() << ";\n";
+
+      out << SP << "std::vector<size_t> idx(dim);";
+
+      
+      out << SP << "for (size_t i = 0; i < " << outputLength << "; i++) {\n";
+      
+      // write here according to size of shape
+      // in generation code can be done automatically
+      // i0 =  i / s0 ; i1 = (i % s0) / s1 ; i2 = ( (i % s0) % s1 ) / s2 and so on
+      // and we have for the inverse
+      // i = i0 * s0 + i1 * s1 + i2 * s2 + i3 * s3 ....
+
+      // don't need to divide by last stride s[n-1] since it is 1 by definition
+      
+      out << SP << SP << "idx[j] = i;\n";
+      out << SP << SP << "size_t k = 0;\n";
+      out << SP << SP << SP << "for(k=0; k < dim-1; k++){\n";
+      out << SP << SP << SP << "idx[k] = idx[k] %" << outputStrides << "[k];\n";
```

Comment:
outputStrides is a vector of integer values. 
You can define this vector in the generated code by doing at the beginning (before looping on `i`): 
```
out << SP << "std::vector<size_t> outputStrides = {" << outputStrides[0];
for (size_t i = 1; i < outputStrides.size(); i++)
   out << ", " << outputStrides[i];
out << "};";
```

and then later here you do: 
```
out << SP << SP << SP << "idx[k] = idx[k] % outputStrides[k];\n";
 ```